controlplane/kubeadm/internal/cluster.go

```diff
@@ -0,0 +1,383 @@
+	"k8s.io/apimachinery/pkg/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	"sigs.k8s.io/cluster-api/controlplane/kubeadm/internal/etcd"
+	etcdutil "sigs.k8s.io/cluster-api/controlplane/kubeadm/internal/etcd/util"
+	"sigs.k8s.io/cluster-api/controlplane/kubeadm/internal/proxy"
+	"sigs.k8s.io/cluster-api/util/certs"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ManagementCluster holds operations on the ManagementCluster
+type ManagementCluster struct {
+	Client ctrlclient.Client
+}
+
+// OwnedControlPlaneMachines returns a MachineFilter function to find all owned control plane machines.
+// Usage: managementCluster.GetMachinesForCluster(ctx, cluster, OwnedControlPlaneMachines(controlPlane.Name))
+func OwnedControlPlaneMachines(controlPlaneName string) func(machine clusterv1.Machine) bool {
+	return func(machine clusterv1.Machine) bool {
+		controllerRef := metav1.GetControllerOf(&machine)
+		if controllerRef == nil {
+			return false
+		}
+		return controllerRef.Kind == "KubeadmControlPlane" && controllerRef.Name == controlPlaneName
+	}
+}
+
+// GetMachinesForCluster returns a list of machines that can be filtered or not.
+// If no filter is supplied then all machines associated with the target cluster are returned.
+func (m *ManagementCluster) GetMachinesForCluster(ctx context.Context, cluster types.NamespacedName, filters ...func(machine clusterv1.Machine) bool) ([]clusterv1.Machine, error) {
+	selector := map[string]string{
+		clusterv1.ClusterLabelName: cluster.Name,
+	}
+	allMachines := &clusterv1.MachineList{}
+	if err := m.Client.List(ctx, allMachines, client.InNamespace(cluster.Namespace), client.MatchingLabels(selector)); err != nil {
+		return nil, errors.Wrap(err, "failed to list machines")
+	}
+	if len(filters) == 0 {
+		return allMachines.Items, nil
+	}
+	filteredMachines := []clusterv1.Machine{}
+	for _, machine := range allMachines.Items {
+		add := true
+		for _, filter := range filters {
+			if !filter(machine) {
+				add = false
+				break
+			}
+		}
+		if add {
+			filteredMachines = append(filteredMachines, machine)
+		}
+	}
+	return filteredMachines, nil
+}
+
+// getCluster builds a cluster object.
+// The cluster is also populated with secrets stored on the management cluster that is required for
+// secure internal pod connections.
+func (m *ManagementCluster) getCluster(ctx context.Context, clusterKey types.NamespacedName) (*cluster, error) {
+	// This adapter is for interop with the `remote` package.
+	adapterCluster := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: clusterKey.Namespace,
+			Name:      clusterKey.Name,
+		},
+	}
+
+	// TODO(chuckha): Unroll remote.NewClusterClient if we are unhappy with getting a restConfig twice.
```

Comment:
I don't know if it's worth filing the issue because there is no action to take/no way to resolve the issue at this time. 

The intent of the comment is to indicate that remote.RESTConfig and remote.NewClusterClient both call remote.RESTConfig and there is an opportunity for clean up if necessary.

I will file an issue for the remote.NewClusterClient to accept a types.NamespacedName: https://github.com/kubernetes-sigs/cluster-api/issues/2304